File: articles/calculus-for-systems-modeling/articles/equilibrium-stability-and-local-dynamics/fortran/equilibrium_stability_audit.f90
program equilibrium_stability_audit
  implicit none
  call print_logistic(0.0d0)
  call print_logistic(100.0d0)
  call print_bistable(0.0d0)
  call print_bistable(0.4d0)
  call print_bistable(1.0d0)
contains
  real(8) function logistic_derivative(x, growth, carrying)
    real(8), intent(in) :: x, growth, carrying
    logistic_derivative = growth*(1.0d0 - 2.0d0*x/carrying)
  end function
  real(8) function bistable_rate(x, threshold)
    real(8), intent(in) :: x, threshold
    bistable_rate = x*(1.0d0-x)*(x-threshold)
  end function
  real(8) function numerical_derivative(x, threshold)
    real(8), intent(in) :: x, threshold
    real(8) :: h
    h = 1.0d-5
    numerical_derivative = (bistable_rate(x+h, threshold) - bistable_rate(x-h, threshold))/(2.0d0*h)
  end function
  character(len=32) function classify(d)
    real(8), intent(in) :: d
    if (d < -1.0d-8) then
      classify = 'locally_stable'
    else if (d > 1.0d-8) then
      classify = 'locally_unstable'
    else
      classify = 'inconclusive_by_linearization'
    end if
  end function
  subroutine print_logistic(eq)
    real(8), intent(in) :: eq
    real(8) :: d
    d = logistic_derivative(eq, 0.6d0, 100.0d0)
    print '(A,F12.6,1X,F12.6,1X,A)', 'logistic_growth', eq, d, trim(classify(d))
  end subroutine
  subroutine print_bistable(eq)
    real(8), intent(in) :: eq
    real(8) :: d
    d = numerical_derivative(eq, 0.4d0)
    print '(A,F12.6,1X,F12.6,1X,A)', 'bistable_threshold', eq, d, trim(classify(d))
  end subroutine
end program

File: articles/calculus-for-systems-modeling/articles/equilibrium-stability-and-local-dynamics/cpp/equilibrium_stability_audit.cpp
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

double logistic_derivative(double x, double growth, double carrying){ return growth*(1.0 - 2.0*x/carrying); }
double bistable_rate(double x, double threshold){ return x*(1.0-x)*(x-threshold); }
double numerical_derivative(double x, double threshold){
  double h = 1e-5;
  return (bistable_rate(x+h, threshold) - bistable_rate(x-h, threshold))/(2.0*h);
}
std::string classify(double d){
  if(d < -1e-8) return "locally_stable";
  if(d > 1e-8) return "locally_unstable";
  return "inconclusive_by_linearization";
}
int main(){
  std::cout << std::fixed << std::setprecision(6);
  std::cout << "scenario,equilibrium,derivative_value,stability,domain_min,domain_max,warning\n";
  for(double eq : std::vector<double>{0.0, 100.0}){
    double d = logistic_derivative(eq, 0.6, 100.0);
    std::cout << "logistic_growth," << eq << "," << d << "," << classify(d) << ",0.000000,100.000000,Logistic stability assumes fixed carrying capacity and smooth density limitation.\n";
  }
  for(double eq : std::vector<double>{0.0, 0.4, 1.0}){
    double d = numerical_derivative(eq, 0.4);
    std::cout << "bistable_threshold," << eq << "," << d << "," << classify(d) << ",0.000000,1.000000,Threshold stability depends on the assumed threshold and domain.\n";
  }
}
